File: fm-domain/src/main/java/io/zeitwert/fm/portfolio/model/ObjPortfolio.java
package io.zeitwert.fm.portfolio.model;

import java.util.Set;

import io.dddrive.core.obj.model.Obj;
import io.zeitwert.fm.account.model.ItemWithAccount;
import io.zeitwert.fm.account.model.ObjAccount;
import io.zeitwert.fm.collaboration.model.ItemWithNotes;
import io.zeitwert.fm.task.model.ItemWithTasks;

public interface ObjPortfolio extends Obj, ItemWithAccount, ItemWithNotes, ItemWithTasks {

	@Override
	Integer getAccountId();

	@Override
	void setAccountId(Integer id);

	@Override
	ObjAccount getAccount();

	String getName();

	void setName(String name);

	String getDescription();

	void setDescription(String description);

	String getPortfolioNr();

	void setPortfolioNr(String portfolioNr);

	Set<Integer> getBuildingSet();

	Set<Integer> getIncludeSet();

	void clearIncludeSet();

	void addInclude(Integer id);

	void removeInclude(Integer id);

	Set<Integer> getExcludeSet();

	void clearExcludeSet();

	void addExclude(Integer id);

	void removeExclude(Integer id);

	double getInflationRate();

	double getDiscountRate();

	Integer getCondition(int year);

	double getPortfolioValue(int year);

}

File: fm-domain/src/main/java/io/zeitwert/fm/portfolio/model/ObjPortfolio.kt
package io.zeitwert.fm.portfolio.model

import dddrive.app.obj.model.Obj
import dddrive.ddd.property.model.AggregateReferenceSetProperty
import io.zeitwert.fm.account.model.ItemWithAccount
import io.zeitwert.fm.building.model.ObjBuilding
import io.zeitwert.fm.collaboration.model.ItemWithNotes
import io.zeitwert.fm.task.model.ItemWithTasks

interface ObjPortfolio :
	Obj,
	ItemWithAccount,
	ItemWithNotes,
	ItemWithTasks {

	var name: String?

	var description: String?

	var portfolioNr: String?

	val buildingSet: AggregateReferenceSetProperty<ObjBuilding>

	val includeSet: AggregateReferenceSetProperty<Obj>

	val excludeSet: AggregateReferenceSetProperty<Obj>

	val inflationRate: Double

	val discountRate: Double

	fun getCondition(year: Int): Int

	fun getPortfolioValue(year: Int): Double
}
